### PR TITLE
planner:dont push down right condition for anti semi join

### PR DESCRIPTION
**Pull request by jingyugao**
### What problem does this PR solve? 
TiDB will push down right condition for anti semi join.
But if we push down the right condition, we can't know why there is no matched rows, null or false.
Fix https://github.com/pingcap/tidb/issues/12074.

### What is changed and how it works?

For anti semi join don't push down the righ condition.
### Check List 

Tests 

 - Unit test

Code changes

Simple change.

Side effects

After this pr,we can't push down the right condition for anti semi join.
So this might decrease the performance.

Related changes

Maybe not

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: planner/core/rule_predicate_push_down.go [hunk 1]
```diff
@@ -173,6 +173,12 @@ func (p *LogicalJoin) PredicatePushDown(predicates []expression.Expression) (ret
 		p.OtherConditions = otherCond
 		leftCond = leftPushCond
 		rightCond = rightPushCond
+		if p.JoinType == AntiSemiJoin {
```
Comment (francis0407) [inline planner/core/rule_predicate_push_down.go]: I think this might be too strict. Look at the explain-test(tpc-h q21)this PR has affected. 
```sql
select
	s_name,
	count(*) as numwait
from
	supplier,
	lineitem l1,
	orders,
	nation
where
	s_suppkey = l1.l_suppkey
	and o_orderkey = l1.l_orderkey
	and o_orderstatus = 'F'
	and l1.l_receiptdate > l1.l_commitdate
	and exists (
		select
			*
		from
			lineitem l2
		where
			l2.l_orderkey = l1.l_orderkey
			and l2.l_suppkey <> l1.l_suppkey
	)
	and not exists (
		select
			*
		from
			lineitem l3
		where
			l3.l_orderkey = l1.l_orderkey
			and l3.l_suppkey <> l1.l_suppkey
			and l3.l_receiptdate > l3.l_commitdate
	)
	and s_nationkey = n_nationkey
	and n_name = 'EGYPT'
group by
	s_name
```
After this change, the subquery condition `l3.l_receiptdate > l3.l_commitdate` cannot be pushed down. This can casue a performance regression. 

Actually, although `not exists` and `not in` both use `anti-semi-join`, but they are not totally equal. I think we'd better fix this issue for `not in`, and not affect the behavior of `not exists`. 

Here are some reference you can take a look:
This is how mysql deals with `exists strategy`.
https://dev.mysql.com/doc/refman/8.0/en/subquery-optimization-with-exists.html

And this is how SparkSQL uses a `null aware anti join` for `not in`(I prefer this method to rewrite the `not in` condition). https://databricks-prod-cloudfront.cloud.databricks.com/public/4027ec902e239c93eaaa8714f173bcfc/2728434780191932/1483312212640900/6987336228780374/latest.html

Comment (francis0407) [reply planner/core/rule_predicate_push_down.go]: Here is how `null aware anti join` works (quoted from Apache Spark):
```
Expand the NOT IN expression with the NULL-aware semantic
to its full form. That is from:
  (a1,a2,...) = (b1,b2,...)
to
  (a1=b1 OR isnull(a1=b1)) AND (a2=b2 OR isnull(a2=b2)) AND ...
```

Comment (jingyugao) [reply planner/core/rule_predicate_push_down.go]: Thanks，maybe I need to rehink about it. Please remove the review request.This commit is too hasty.

Comment (eurekaka) [reply planner/core/rule_predicate_push_down.go]: You can check the `InOperand` field of `RightConditions`'s arguments to decide if we should disable pushdown for it. `InOperand` is introduced to differentiate the `AntiSemiJoin` / `LeftAntiSemiJoin` / `AntiLeftOuterSemiJoin` generated by `NOT IN` and `NOT EXISTS`.

File: cmd/explaintest/r/explain_easy.result [hunk 2]
```diff
@@ -226,22 +226,20 @@ StreamAgg_12	1.00	root	funcs:sum(col_0)
...
+  └─TableReader_13	10000.00	root	data:TableScan_12
+    └─TableScan_12	10000.00	cop	table:t2, range:[-inf,+inf], keep order:false, stats:pseudo
 explain select sum(6 in (select c2 from t2)) from t1;
 id	count	task	operator info
 StreamAgg_12	1.00	root	funcs:sum(col_0)
-└─Projection_22	10000.00	root	cast(5_aux_0)
-  └─HashLeftJoin_21	10000.00	root	CARTESIAN left outer semi join, inner:TableReader_20
+└─Projection_21	10000.00	root	cast(5_aux_0)
+  └─HashLeftJoin_20	10000.00	root	CARTESIAN left outer semi join, inner:TableReader_19, other cond:eq(6, test.t2.c2)
     ├─IndexReader_17	10000.00	root	index:IndexScan_16
```
Comment (jingyugao) [inline cmd/explaintest/r/explain_easy.result]: here the `eq(6, test.t2.c2)` is moved into other condition because it is from `6 in (subq)`

File: cmd/explaintest/r/explain_easy.result [hunk 3]
```diff
@@ -281,22 +279,22 @@ color=black
...
 color=black
 label = "cop"
 "IndexScan_10"
 }
-subgraph cluster13{
+subgraph cluster12{
 node [style=filled, color=lightgrey]
 color=black
 label = "cop"
-"Selection_13" -> "TableScan_12"
```
Comment (jingyugao) [inline cmd/explaintest/r/explain_easy.result]: `Selection_13` is `eq(1,t2.c2)` which is from `1 in (subq)`

File: cmd/explaintest/r/explain_easy_stats.result [hunk 4]
```diff
@@ -123,12 +123,11 @@ set @@session.tidb_opt_insubq_to_join_and_agg=0;
 explain select 1 in (select c2 from t2) from t1;
 id	count	task	operator info
 Projection_6	1999.00	root	5_aux_0
-└─HashLeftJoin_7	1999.00	root	CARTESIAN left outer semi join, inner:TableReader_14
+└─HashLeftJoin_7	1999.00	root	CARTESIAN left outer semi join, inner:TableReader_13, other cond:eq(1, test.t2.c2)
```
Comment (jingyugao) [inline cmd/explaintest/r/explain_easy_stats.result]: `eq(1, test.t2.c2)` is from `1 in (subq)`

File: cmd/explaintest/r/explain_easy_stats.result [hunk 5]
```diff
@@ -139,22 +138,22 @@ color=black
...
 color=black
 label = "cop"
 "IndexScan_10"
 }
-subgraph cluster13{
+subgraph cluster12{
 node [style=filled, color=lightgrey]
 color=black
 label = "cop"
-"Selection_13" -> "TableScan_12"
```
Comment (jingyugao) [inline cmd/explaintest/r/explain_easy_stats.result]: `Selection_13` is from `1 in (subq)`

File: executor/join_test.go [hunk 6]
```diff
@@ -787,6 +787,15 @@ func (s *testSuite2) TestSubquery(c *C) {
...
 	tk.MustQuery("select * from t1 where a in (select a from t2)").Check(testkit.Rows("1"))
 
+	tk.MustExec("insert into t2 value(null)")
+	tk.MustQuery("select * from t1 where 1 in (select b from t2)").Check(testkit.Rows("1"))
+	tk.MustQuery("select * from t1 where 1 not in (select b from t2)").Check(testkit.Rows())
+	tk.MustExec("delete from t2 where b=1")
+	tk.MustQuery("select 1 in (select b from t2) from t1").Check(testkit.Rows("<nil>"))
+	tk.MustQuery("select 1 not in (select b from t2) from t1").Check(testkit.Rows("<nil>"))
+	// TODO: this query will cause an index out of range panic
+	// tk.MustQuery("select 1 not in (select null from t1) from t2").Check(testkit.Rows())
```
Comment (jingyugao) [inline executor/join_test.go]: this is another bug when `executing physical_plan`,so I can't add this test case.

**Comment by jingyugao** [comment]
PTAL @francis0407  @eurekaka

File: executor/join_test.go [hunk 7]
```diff
@@ -787,6 +787,22 @@ func (s *testSuite2) TestSubquery(c *C) {
 	tk.MustExec("insert into t2 values(1)")
 	tk.MustQuery("select * from t1 where a in (select a from t2)").Check(testkit.Rows("1"))
 
+	tk.MustExec("insert into t2 value(null)")
+	tk.MustQuery("select * from t1 where 1 in (select b from t2)").Check(testkit.Rows("1"))
+	tk.MustQuery("select * from t1 where 1 not in (select b from t2)").Check(testkit.Rows())
```
Comment (francis0407) [inline executor/join_test.go]: How about `select * from t1 where 2 not in (select b from t2)`? This query should return an empty set.

File: executor/join_test.go [hunk 8]
```diff
@@ -787,6 +787,22 @@ func (s *testSuite2) TestSubquery(c *C) {
...
+	tk.MustQuery("select 1 not in (select b from t2) from t1").Check(testkit.Rows("0"))
+	tk.MustQuery("select 1 not in (select 1 from t2) from t1").Check(testkit.Rows("0"))
+
+	tk.MustExec("delete from t2 where b=1")
+	tk.MustQuery("select 1 in (select b from t2) from t1").Check(testkit.Rows("<nil>"))
+	tk.MustQuery("select 1 not in (select b from t2) from t1").Check(testkit.Rows("<nil>"))
+	tk.MustQuery("select 1 not in (select 1 from t2) from t1").Check(testkit.Rows("0"))
+	tk.MustQuery("select 1 in (select 1 from t2) from t1").Check(testkit.Rows("1"))
+	// TODO: this query will cause an index out of range panic
+	// tk.MustQuery("select 1 not in (select null from t1) from t2").Check(testkit.Rows())
```
Comment (francis0407) [inline executor/join_test.go]: Have we checked `null in (select ...)` and `null not in (select ...)` ?

File: expression/util.go [hunk 9]
```diff
@@ -26,7 +26,7 @@ import (
 	"github.com/pingcap/parser/terror"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
-	"github.com/pingcap/tidb/types/parser_driver"
+	driver "github.com/pingcap/tidb/types/parser_driver"
```
Comment (francis0407) [inline expression/util.go]: It seems that this change has nothing to do with this PR?

File: expression/expression.go [hunk 10]
```diff
@@ -165,9 +165,7 @@ func IsEQCondFromIn(expr Expression) bool {
 	if !ok || sf.FuncName.L != ast.EQ {
 		return false
 	}
-	cols := make([]*Column, 0, 1)
-	cols = ExtractColumnsFromExpressions(cols, sf.GetArgs(), isColumnInOperand)
```
Comment (francis0407) [inline expression/expression.go]: We can remove the function `IsColumnInOperand` now.

File: executor/join_test.go (hunk 8 above)
Comment (jingyugao) [reply executor/join_test.go]: No,because this will casue a panic when execute join(hash join,SMJ,NLJ. all).I will try to fix it later.

Comment (francis0407) [reply executor/join_test.go]: I test the query before this PR:
```
mysql> select 1 in (select null from t1) from t;
+----------------------------+
| 1 in (select null from t1) |
+----------------------------+
|                          0 |
+----------------------------+
1 row in set (0.00 sec)

mysql> select 1 not in (select null from t1) from t;
+--------------------------------+
| 1 not in (select null from t1) |
+--------------------------------+
|                              1 |
+--------------------------------+
1 row in set (0.01 sec)
```
Is the panic related to this PR?

File: expression/expression.go (hunk 10 above)
Comment (jingyugao) [reply expression/expression.go]: removed

File: expression/util.go (hunk 9 above)
Comment (jingyugao) [reply expression/util.go]: fixed

File: executor/join_test.go (hunk 7 above)
Comment (jingyugao) [reply executor/join_test.go]: test case added

File: executor/join_test.go (hunk 8 above)
Comment (jingyugao) [reply executor/join_test.go]: Try this sql `select a in (select 0)xx from (select null as a)x`. This will cause panic in master because join can't hadler 'other cond' correctly. 
After this pr,this sql will generate a 'other cond'.

File: planner/core/expression_rewriter.go [hunk 11]
```diff
@@ -712,21 +715,35 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
...
-	if np.Schema().Len() == 1 {
-		rexpr = np.Schema().Columns[0]
-		rCol := rexpr.(*expression.Column)
-		// For AntiSemiJoin/LeftOuterSemiJoin/AntiLeftOuterSemiJoin, we cannot treat `in` expression as
-		// normal column equal condition, so we specially mark the inner operand here.
-		if v.Not || asScalar {
+	// For AntiSemiJoin/LeftOuterSemiJoin/AntiLeftOuterSemiJoin, we cannot treat `in` expression as
+	// normal column equal condition, so we specially mark the inner operand here.
+	if v.Not || asScalar {
+		for i, rCol := range np.Schema().Columns {
```
Comment (francis0407) [inline planner/core/expression_rewriter.go]: This change also extends this method to handle `multi-column IN`? If so, we'd better add tests like `(1, null) in (...)`, `(a, null) in (...)` ....

File: planner/core/expression_rewriter.go [hunk 12]
```diff
@@ -712,21 +715,35 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
...
+			lexpr2 := lexpr
+			lSf, isSf := lexpr.(*expression.ScalarFunction)
+			if isSf {
+				lexpr2 = lSf.GetArgs()[i]
```
Comment (francis0407) [inline planner/core/expression_rewriter.go]: The name `lexpr2` is strange, can we use other name?

File: planner/core/logical_plan_builder.go [hunk 13]
```diff
@@ -283,6 +283,12 @@ func (p *LogicalJoin) extractOnCondition(conditions []expression.Expression, der
 					continue
 				}
 			}
+
+			if binop.FuncName.L == ast.EQ && expression.IsEQCondFromIn(expr) {
+				// should I check null flag?
```
Comment (francis0407) [inline planner/core/logical_plan_builder.go]: I'm not sure about this. cc @eurekaka

File: util/chunk/mutrow.go [hunk 14]
```diff
@@ -79,7 +79,7 @@ func zeroValForType(tp *types.FieldType) interface{} {
 			return uint64(0)
 		}
 		return int64(0)
-	case mysql.TypeString, mysql.TypeVarString, mysql.TypeVarchar:
+	case mysql.TypeString, mysql.TypeVarString, mysql.TypeVarchar, mysql.TypeNull:
```
[the remark this answers is not in the document]
Comment (jingyugao) [reply util/chunk/mutrow.go]: Yes,in master `null` will be treated as fix length type.
Then in `ShallowCopyPartialRow`,  the `dst.offset[1]` will cause a panic.

File: planner/core/expression_rewriter.go [hunk 15]
```diff
@@ -349,6 +349,9 @@ func (er *expressionRewriter) buildSemiApplyFromEqualSubq(np LogicalPlan, l, r e
 		if lCol, ok := l.(*expression.Column); ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
 			rCol.InOperand = false
 		}
+		if lcon, ok := l.(*expression.Constant); ok && !lcon.Value.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
```
[the remark this answers is not in the document]
Comment (jingyugao) [reply planner/core/expression_rewriter.go]: fixed

File: planner/core/expression_rewriter.go [hunk 16]
```diff
@@ -712,21 +715,35 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
...
+			}
 			rCol.InOperand = true
 			// If both input columns of `in` expression are not null, we can treat the expression
 			// as normal column equal condition instead.
-			lCol, ok := lexpr.(*expression.Column)
+			lCol, ok := lexpr2.(*expression.Column)
 			if ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
 				rCol.InOperand = false
 			}
+			lcon, ok := lexpr2.(*expression.Constant)
```
[the remark this answers is not in the document]
Comment (jingyugao) [reply planner/core/expression_rewriter.go]: fixed

File: util/chunk/mutrow.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (eurekaka) [reply util/chunk/mutrow.go]: @jingyugao Could you please explain more about this change?

File: planner/core/logical_plan_builder.go (hunk 13 above)
Comment (eurekaka) [reply planner/core/logical_plan_builder.go]: We don't need to check Null flag here now, since we have checked it already when marking `InOperand`. Can we move this branch to the front of the for loop and remove the `InOperand` check in line https://github.com/pingcap/tidb/pull/12075/files#diff-b9e75a99170ea35a67074d5d3d3305adR280?

File: expression/constant.go [hunk 17]
```diff
@@ -60,6 +60,9 @@ type Constant struct {
 	// It's only used to reference a user variable provided in the `EXECUTE` statement or `COM_EXECUTE` binary protocol.
 	ParamMarker *ParamMarker
 	hashcode    []byte
+	// InOperand indicates whether this column is the inner operand of column equal condition converted
+	// from `[not] in (subq)`.
+	InOperand bool
```
Comment (eurekaka) [inline expression/constant.go]: Why do we need this field in `Constant`?

File: planner/core/expression_rewriter.go [hunk 18]
```diff
@@ -349,6 +349,9 @@ func (er *expressionRewriter) buildSemiApplyFromEqualSubq(np LogicalPlan, l, r e
 		if lCol, ok := l.(*expression.Column); ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
 			rCol.InOperand = false
 		}
+		if lcon, ok := l.(*expression.Constant); ok && !lcon.Value.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+			rCol.InOperand = false
+		}
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]: We should check cases like `CorrelatedColumn in Column` as well?

File: planner/core/expression_rewriter.go [hunk 19]
```diff
@@ -712,21 +715,35 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
...
 			// as normal column equal condition instead.
-			lCol, ok := lexpr.(*expression.Column)
+			lCol, ok := lexpr2.(*expression.Column)
 			if ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
 				rCol.InOperand = false
 			}
+			lcon, ok := lexpr2.(*expression.Constant)
+			if ok && !lcon.Value.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+				rCol.InOperand = false
+			}
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]: Ditto, check cases like `CorrelatedColumn in Column` here and reset `InOperand` if we can?

File: util/chunk/mutrow.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (eurekaka) [reply util/chunk/mutrow.go]: Shall we fix it in `ShallowCopyPartialRow`? It looks dangerous that zero value for `TypeNull` is an empty string.

[the remark this answers is not in the document]
Comment (eurekaka) [reply util/chunk/mutrow.go]: @zz-jason @SunRunAway Please take a look at this change.

[the remark this answers is not in the document]
Comment (zz-jason) [reply util/chunk/mutrow.go]: for fixed-length types, we should store it in the same way as `bigint`/`double`.

**Comment by eurekaka** [comment]
@jingyugao Please resolve conflicts and fold all the comments resolved, thanks.

File: planner/core/expression_rewriter.go [hunk 20]
```diff
@@ -712,21 +715,35 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
...
+			if isSf {
+				lSf.GetArgs()[i] = lexpr2
```
[the remark this answers is not in the document]
Comment (jingyugao) [reply planner/core/expression_rewriter.go]: rmoved

File: expression/constant.go (hunk 17 above)
Comment (jingyugao) [reply expression/constant.go]: ```select a not in (select 1 )``` like this

File: planner/core/expression_rewriter.go (hunk 18 above)
Comment (jingyugao) [reply planner/core/expression_rewriter.go]: Yes, added

File: planner/core/expression_rewriter.go (hunk 19 above)
Comment (jingyugao) [reply planner/core/expression_rewriter.go]: added

File: planner/core/expression_rewriter.go (hunk 12 above)
Comment (jingyugao) [reply planner/core/expression_rewriter.go]: changed to larg.

File: planner/core/expression_rewriter.go [hunk 21]
```diff
@@ -400,6 +400,13 @@ func (er *expressionRewriter) buildSemiApplyFromEqualSubq(np LogicalPlan, l, r e
 		if lCol, ok := l.(*expression.Column); ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
 			rCol.InOperand = false
 		}
+		if lCon, ok := l.(*expression.Constant); ok && !lCon.Value.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+			rCol.InOperand = false
+		}
+		lClc, ok := l.(*expression.CorrelatedColumn)
+		if ok && !lClc.Data.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]: `Data` is filled in execution, it has not been initialized here. I think we should check whether `Column` of correlated column has NotNull flag instead.

How about:
```golang
corCol, ok := l.(*expression.CorrelatedColumn)
if ok && mysql.HasNotNullFlag(corCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
```

File: planner/core/expression_rewriter.go (hunk 11 above)
Comment (eurekaka) [reply planner/core/expression_rewriter.go]: +1

File: planner/core/expression_rewriter.go [hunk 22]
```diff
@@ -762,21 +769,36 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
...
 				rCol.InOperand = false
 			}
+			lCon, ok := larg.(*expression.Constant)
+			if ok && !lCon.Value.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+				rCol.InOperand = false
+			}
+			lClc, ok := larg.(*expression.CorrelatedColumn)
+			if ok && !lClc.Data.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+				rCol.InOperand = false
+			}
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]: Extract this part of logic as a single function, and reuse it in `buildSemiApplyFromEqualSubq` as well?

File: tools/check/check-gogenerate.sh [hunk 23]
```diff
@@ -5,7 +5,7 @@ go generate ./...
 set +e
 diffline=$(git status -s | awk '{print $2}' | xargs grep '^// Code generated .* DO NOT EDIT\.$' 2>/dev/null | wc -l)
 set -e
-if [[ $diffline != 0 ]]
+if [[ $diffline > 0 ]]
```
Comment (eurekaka) [inline tools/check/check-gogenerate.sh]: Unnecessary change?

File: tools/check/check-gogenerate.sh [hunk 24]
```diff
@@ -5,7 +5,7 @@ go generate ./...
 set +e
 diffline=$(git status -s | awk '{print $2}' | xargs grep '^// Code generated .* DO NOT EDIT\.$' 2>/dev/null | wc -l)
 set -e
-if [[ $diffline != 0 ]]
+if [[ $diffline -ne 0 ]]
```
Comment (jingyugao) [inline tools/check/check-gogenerate.sh]: https://stackoverflow.com/questions/20449543/bash-equality-operators-eq

File: planner/core/logical_plan_builder.go [hunk 25]
```diff
@@ -299,7 +303,7 @@ func (p *LogicalJoin) extractOnCondition(conditions []expression.Expression, der
 					// false even if t.a is null or s.a is null. To make this join "empty aware",
 					// we should differentiate `t.a = s.a` from other column equal conditions, so
 					// we put it into OtherConditions instead of EqualConditions of join.
-					if binop.FuncName.L == ast.EQ && !arg0.InOperand && !arg1.InOperand {
+					if binop.FuncName.L == ast.EQ {
```
Comment (eurekaka) [inline planner/core/logical_plan_builder.go]: Please move the comments here to the front as well.

File: expression/expression.go [hunk 26]
```diff
@@ -158,19 +158,13 @@ func (e CNFExprs) Shallow() CNFExprs {
...
 // IsEQCondFromIn checks if an expression is equal condition converted from `[not] in (subq)`.
 func IsEQCondFromIn(expr Expression) bool {
 	sf, ok := expr.(*ScalarFunction)
 	if !ok || sf.FuncName.L != ast.EQ {
 		return false
 	}
-	cols := make([]*Column, 0, 1)
-	cols = ExtractColumnsFromExpressions(cols, sf.GetArgs(), isColumnInOperand)
-	return len(cols) > 0
+	return exprsContainInOperand(sf.GetArgs())
```
Comment (eurekaka) [inline expression/expression.go]: This change is unnecessary now?

File: expression/util.go [hunk 27]
```diff
@@ -114,6 +114,22 @@ func ExtractColumns(expr Expression) []*Column {
 	return extractColumns(result, expr, nil)
 }
 
+func exprsContainInOperand(exprs []Expression) bool {
```
Comment (eurekaka) [inline expression/util.go]: ditto

**Comment by winoros** [comment]
Some comments not about the main changes.
Since the special condition cannot be pushed down to the join's children.
We can let the join have one more field like `NullSensitiveConditions` and put them into this field.
Then the `InOperand` field can be removed from `expression.Column`?

File: expression/expression.go [hunk 28]
```diff
@@ -173,6 +173,20 @@ func IsEQCondFromIn(expr Expression) bool {
 	return len(cols) > 0
 }
 
+// NotNull checks if an expression is constantly not null.
+func NotNull(l Expression) bool {
```
Comment (fzhedu) [inline expression/expression.go]: This function would be more powerful. Like binary operators: col op col (const). For example 
```
mysql> create table ss (a int not null, b int);
mysql> create table tt (a int not null, b int);

mysql> explain select  ((t.a,t.b) in (select a,b from ss as s)) from tt as t;
+-------------------------+----------+-----------+---------------------------------------------------------------------------------------------------------------+
| id                      | count    | task      | operator info                                                                                                 |
+-------------------------+----------+-----------+---------------------------------------------------------------------------------------------------------------+
| Projection_6            | 10000.00 | root      | Column#9                                                                                                      |
| └─HashLeftJoin_7        | 10000.00 | root      | left outer semi join, inner:TableReader_11, equal:[eq(Column#1, Column#4)], other cond:eq(Column#2, Column#5) |
|   ├─TableReader_9       | 10000.00 | root      | data:TableScan_8                                                                                              |
|   │ └─TableScan_8       | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                                                    |
|   └─TableReader_11      | 10000.00 | root      | data:TableScan_10                                                                                             |
|     └─TableScan_10      | 10000.00 | cop[tikv] | table:s, range:[-inf,+inf], keep order:false, stats:pseudo                                                    |
+-------------------------+----------+-----------+---------------------------------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)

mysql> explain select  ((t.a+1,t.b) in (select a,b from ss as s)) from tt as t;
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------------------+
| id                      | count    | task      | operator info                                                                                                            |
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------------------+
| Projection_6            | 10000.00 | root      | Column#9                                                                                                                 |
| └─HashLeftJoin_7        | 10000.00 | root      | CARTESIAN left outer semi join, inner:TableReader_11, other cond:eq(Column#2, Column#5), eq(plus(Column#1, 1), Column#4) |
|   ├─TableReader_9       | 10000.00 | root      | data:TableScan_8                                                                                                         |
|   │ └─TableScan_8       | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                                                               |
|   └─TableReader_11      | 10000.00 | root      | data:TableScan_10                                                                                                        |
|     └─TableScan_10      | 10000.00 | cop[tikv] | table:s, range:[-inf,+inf], keep order:false, stats:pseudo                                                               |
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)
```
These second can avoid the CARTESIAN join.

File: planner/core/expression_rewriter.go [hunk 29]
```diff
@@ -762,21 +762,27 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
...
+				larg = lSf.GetArgs()[i]
+			}
 			rCol.InOperand = true
-			// If both input columns of `in` expression are not null, we can treat the expression
-			// as normal column equal condition instead.
-			lCol, ok := lexpr.(*expression.Column)
-			if ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+			// If both input columns of `!= all / = any` expression are not null, we can treat the expression
+			// as normal column equal condition.
+			if mysql.HasNotNullFlag(rCol.GetType().Flag) && expression.NotNull(larg) {
```
Comment (fzhedu) [inline planner/core/expression_rewriter.go]: The conditions to get rid of `in` are limited;
```
mysql> explain select  (t.a in (select s.a from s)) **is true** from t;
+-------------------------+----------+-----------+-----------------------------------------------------------------------------------------+
| id                      | count    | task      | operator info                                                                           |
+-------------------------+----------+-----------+-----------------------------------------------------------------------------------------+
| Projection_6            | 10000.00 | root      | istrue(Column#8)                                                                        |
| └─HashLeftJoin_7        | 10000.00 | root      | CARTESIAN left outer semi join, inner:TableReader_11, other cond:eq(Column#1, Column#4) |
|   ├─TableReader_9       | 10000.00 | root      | data:TableScan_8                                                                        |
|   │ └─TableScan_8       | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                              |
|   └─TableReader_11      | 3.00     | root      | data:TableScan_10                                                                       |
|     └─TableScan_10      | 3.00     | cop[tikv] | table:s, range:[-inf,+inf], keep order:false, stats:pseudo                              |
+-------------------------+----------+-----------+-----------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)
```
() is true above the subquery can also get rid of `in`, so that it can be converted to a inner join.

File: expression/expression.go (hunk 28 above)
Comment (jingyugao) [reply expression/expression.go]: Yes,I should consider scalar function.

File: planner/core/expression_rewriter.go (hunk 29 above)
Comment (jingyugao) [reply planner/core/expression_rewriter.go]: After considering scalar function, `eq(Column#1, Column#4)`  is treated as euqal condition now.But how can it be converted to inner join?

**Comment by jingyugao** [comment]
> Since the special condition cannot be pushed down to the join's children.

Modify the joiner? I think this is the right way. Current way seem to put a patch.

**Comment by eurekaka** [comment]
> Some comments not about the main changes.
> Since the special condition cannot be pushed down to the join's children.
> We can let the join have one more field like `NullSensitiveConditions` and put them into this field.
> Then the `InOperand` field can be removed from `expression.Column`?
> @eurekaka @lamxTyler @francis0407

Yes, code would be cleaner then. We can do it after this PR get merged.

File: planner/core/expression_rewriter.go (hunk 29 above)
Comment (fzhedu) [reply planner/core/expression_rewriter.go]: I find that `() is true` is the only special case where the `in subquery` can be converted to `exists subquery` . In other most cases, the `in subquery` should return null, false or true in the projection clause.
So we may ignore the optimization for `() is true` here.

**Comment by fzhedu** [comment]
I find the converted conditions from `in subquery` in semi join are pushed down. I wonder whether this would induce errors?
```
mysql> explain select * from s where (2 in (select a from t where (s.b=t.b) is null));
+------------------------+-------+-----------+--------------------------------------------------------------------------------------+
| id                     | count | task      | operator info                                                                        |
+------------------------+-------+-----------+--------------------------------------------------------------------------------------+
| HashLeftJoin_9         | 4.00  | root      | CARTESIAN semi join, inner:TableReader_14, other cond:isnull(eq(Column#2, Column#5)) |
| ├─TableReader_11       | 5.00  | root      | data:TableScan_10                                                                    |
| │ └─TableScan_10       | 5.00  | cop[tikv] | table:s, range:[-inf,+inf], keep order:false, stats:pseudo                           |
| └─TableReader_14       | 0.01  | root      | data:Selection_13                                                                    |
|   └─Selection_13       | 0.01  | cop[tikv] | eq(2, Column#4)                                                                      |
|     └─TableScan_12     | 5.00  | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                           |
+------------------------+-------+-----------+--------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)
```

**Comment by fzhedu** [comment]
> > Some comments not about the main changes.
> > Since the special condition cannot be pushed down to the join's children.
> > We can let the join have one more field like `NullSensitiveConditions` and put them into this field.
> > Then the `InOperand` field can be removed from `expression.Column`?
> > @eurekaka @lamxTyler @francis0407
> 
> Yes, code would be cleaner then. We can do it after this PR get merged.

I also got the same way to store the null-aware conditions, and they cannot be pushed down. Adding a `nullAwareConditions` can also solve the bugs solved by this PR. But some conditions  in `nullAwareConditions` can also be pushed down, like the referred columns are not null. So it also needs to add the `func NotNull(l Expression) bool {}` functions in this PR to push down the non-null conditions.

**Comment by winoros** [comment]
@fzhedu 
When it's not null, `inOperand` is also false currently. This won't generate new cases which need to be special handled.

File: expression/expression.go [hunk 30]
```diff
@@ -173,6 +173,27 @@ func IsEQCondFromIn(expr Expression) bool {
...
+	switch l := l.(type) {
+	case *Column:
+		return mysql.HasNotNullFlag(l.GetType().Flag)
+	case *Constant:
+		return !l.Value.IsNull()
+	case *CorrelatedColumn:
+		return mysql.HasNotNullFlag(l.GetType().Flag)
+	case *ScalarFunction:
+		for _, arg := range l.GetArgs() {
```
Comment (winoros) [inline expression/expression.go]: Though the current implementation may satisfy the `InSubquery`'s needs. But for the real `NotNull` property of scalar function, we need to take the `IsNull` sig in to special consideration at least.

**Comment by fzhedu** [comment]
> @fzhedu
> When it's not null, `inOperand` is also false currently. This won't generate new cases which need to be special handled.

Yes. The way in this PR and adding a `nullAwareConditions` are two ways to solve the current bugs about null in subquery. Such two ways both needs to push down the non-null conditions converted from `in subqueries`. This is done by setting `inOperand` in this PR. It is better to take the second way? @winoros

File: expression/expression.go (hunk 30 above)
Comment (jingyugao) [reply expression/expression.go]: yes, for example `arg is true` is not null when arg is null.

**Comment by jingyugao** [comment]
> explain select * from s where (2 in (select a from t where (s.b=t.b) is null));

This pr seems not change this behavior. Execute plan is same between master and this pr.

File: planner/core/expression_rewriter.go (hunk 22 above)
Comment (jingyugao) [reply planner/core/expression_rewriter.go]: fixed

File: planner/core/logical_plan_builder.go (hunk 25 above)
Comment (jingyugao) [reply planner/core/logical_plan_builder.go]: fixed

**Comment by fzhedu** [comment]
I found the PR also has errors, please see https://github.com/pingcap/tidb/issues/13743
